synapse/logging/_remote.py

```diff
@@ -146,17 +147,21 @@ def _connect(self) -> None:
 
         self._connection_waiter = self._service.whenConnected(failAfterFailures=1)
 
-        @self._connection_waiter.addErrback
-        def fail(r):
-            r.printTraceback(file=sys.__stderr__)
+        def fail(failure: Failure) -> None:
+            # If the Deferred was cancelled (e.g. during shutdown) do not try to
+            # reconnect (this will cause an infinite loop of errors).
+            if failure.check(CancelledError):
+                return
```

Comment:
This actual fix is returning early if a `CancelledError` is received. The rest is just a bit of house-cleaning:
* Avoid single-character variables.
* Using `addCallbacks` instead of the decorators, which I find a bit more readable.
* Type-hints.

Reply:
I think there are other situations where we might get a cancelled error? Should we check for a cancelled error *and* if we're shutting down?

Reply:
I think we would want to avoid re-connecting anytime that we got a cancelled error, e.g. if the logging configuration changed and `handler.close()` is called? Is there another case you're thinking of?

Reply:
Timeouts and things sometime bubble up as cancelled errors, maybe? I'm just nervous that if we do get a cancelled error spuriously we'll just stop sending logs entirely?

Reply:
Ah, interesting. Adding a timeout to a `Deferred` ends up calling `cancel()` on it, but adds a callback which automatically transfers that to a `TimeOutError`.

The [`whenConnected`](https://twistedmatrix.com/documents/current/api/twisted.application.internet.ClientService.html#whenConnected) docs say that it will be raised when the "service is stopped".

We can add a flag though. 😄

Reply:
@erikjohnston I think 315d45de3a69ffe1253b8609c50e853f9f402d98 will assuage your concerns.

Reply:
Thanks! I'm not sure I'd trust the twisted docs enough to rely on that behaviour :)